AcademyApp/ViewControllers/LoginViewController.swift

```diff
@@ -0,0 +1,189 @@
+        doneButton.isEnabled = isValidEmail(possibleEmail: emailTextField.text!) && passwordTextField.text!.count > 0 ? true : false
+    }
+    
+    // control event : touch down (hold)
+    @IBAction func showPassword(_ sender: UIButton) {
+        passwordTextField.isSecureTextEntry = false
+    }
+    
+    // control event : touch up inside (release)
+    @IBAction func hidePassword(_ sender: UIButton) {
+        passwordTextField.isSecureTextEntry = true
+    }
+}
+
+// MARK - UITextField's delegate methods
+extension LoginViewController : UITextFieldDelegate
+{
+    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
+        self.view.endEditing(true)
+        return true
+    }
+}
+
+// MARK: - Responding to keyboard notifications
+extension LoginViewController
+{
+    private func addObservers(){
+        NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillShow(notification:)), name: UIResponder.keyboardWillShowNotification, object: nil)
+        NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillHide(notification:)), name: UIResponder.keyboardWillHideNotification, object: nil)
+    }
+    
+    private func removeObservers(){
+        NotificationCenter.default.removeObserver(self, name: UIResponder.keyboardWillShowNotification, object: nil)
+        NotificationCenter.default.removeObserver(self, name: UIResponder.keyboardWillHideNotification, object: nil)
+    }
+    
+    @objc func keyboardWillShow(notification: Notification){
+            
+        guard let keyboardRect = (notification.userInfo?[UIResponder.keyboardFrameEndUserInfoKey] as? NSValue)?.cgRectValue  else {
+            return
+        }
+            
+        let keyboardHeight = keyboardRect.height
+        doneButtonBottomConstraint.constant =  keyboardHeight
+        UIView.animate(withDuration: 0.5) {
+            self.view.layoutIfNeeded()
+        }
+    }
+    
+    @objc func keyboardWillHide(notification: Notification){
+        
+        doneButtonBottomConstraint.constant =  defaultDoneButtonBottomSpacing
+        UIView.animate(withDuration: 0.5) {
+            self.view.layoutIfNeeded()
+        }
+    }
+}
+
+// MARK: - UI Setup Methods
+extension LoginViewController
+{
+    var textFields : [UITextField] {
+        return [emailTextField, passwordTextField]
+    }
+    
+    private func setupUI(){
+        backgroundColorSetup()
+        textFieldsSetup()
+        showLabelSetup()
+        horizontalLinesSetup()
+        doneButtonSetup()
+    }
+    
+    private func backgroundColorSetup(){
+        view.backgroundColor = UIColor.almostBlack
+    }
+    
+    private func horizontalLinesSetup(){
+        horizontalLineViews.forEach {
+            $0.backgroundColor = UIColor.blackish
+        }
+    }
+    
+    private func textFieldsSetup(){
+        
+        let fontName = "MaisonNeue"
+        let font = UIFont(name: fontName, size: 16) ?? UIFont(name: "Helvetica", size: 16)!
```

Comment:
Is this similar to working with colors, isnt it? What about UIFont extension?

Reply:
Moved fonts to UIFontExtensions